LoRaWAN: ClockSynch/Multicast/Fragmentation Plugins & session handling

hasnainvirk opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
This PR provides implementation for three different protocol specs:
i) LoRaWAN  Application Layer Clock Synchronization Specification v1.0.0
ii) LoRaWAN Remote Multicast Setup Specification v1.0.0
iii) LoRaWAN Fragmented Data Block Transport Specification v1.0.0
The implementations of these protocols are provided as plug-ins/add-ins/application-extensions and can be used by the application on-demand. The plug-ins can be used independently as well as in association to perform various tasks, e.g., device synchronization, multicast sessions, transport of larger data in fragments, firmware updates etc. 

In addition to that a bunch of new APIs are added which serve multiple purposes. They fill-in the holes in our current offering plus they facilitate the plug-ins.

One of the classes introduced here, FragAssembler is directly derived from a 3rd party source and is used for FEC. The original license is maintained, and augmented with Arm BSD-Clause3 license. Only style and some minor performance tweaks are added to the original source.   

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@AnttiKauppila @janjongboom 
<!-- 
    Optional
    Request additional reviewers with @username
-->

Commit:
Adding ability to change/restore RX2 frequency

Usually RX2 frequency is fixed as per regional parameter specifications
but there can be exceptions. Network servers are allowed to change RX2
frequency using Mac commands or out of the band mechanisms. We have
introduces APIs in this commit which will enable the application to
modify or restore RX2 window frequency.

Commit:
Adding Multicast register & multicast reception

The stack will now maintain a register with a maximum of 4 multicast
session entries. The maximum number of session contexts is dictated by
the LoRaWAN remote multicast setup specifcation v1.0.0.

'register_multicast_address()' API is used to set an entry into the
multicast register, 'get_multicast_addr_register()' API is used to acquire a
handle to the multicast register itself and
'verify_multicast_freq_and_dr()' API is used to verify frequency and
data rate parameters if requested by the network in a special case
(i.e., a switch to class C at a given time with a given frequency and SF).

When a multicast downlink is received, the stack looks-up the multicast
register for the address match. If a match is found, other session
specific parameters in the register entry, i.e., frmae ocunters and security
keys are loaded to be used for the stack for successful reception and
decryption. The application is notified about the reception as it would
normally be notified (i.e., RX_DONE event).

If the downlink doesn't match any of the multicast addresses or the
unicast address, the packet is dropped and an adrress failure is
notified to the application.

Commit:
Helper bit manipulation functions

Trivial helper functions perform operations on bit fields and masks.

Commit:
Adding ClockSync Plugin

This plug-in or application extension implements LoRaWAN Application
Layer Clock Synchronization Specification v1.0.0. It is supposed to
exist in-tree as an on-demand plugin. Applications can use the plug-in
to synchronize their clocks with network if the protocol version is
v1.0.2 or lower. For v1.0.3 and higher, use of DevTimeReq mac command is
recommended.

The plugin is completely detached from the core stack or application
logic. It just provides easy to use interface to handle Clock sync
commands. The application is reponsible for actual transmission of
uplinks at the clock sync port and evoking the plugin upon reception of
a clock sync command.

Commit:
Adding Multicast control package

This is an implementation of LoRaWAN Remote Multicast Setup
Specification v1.0.0.
It is supposed to be used as an on-demand plug-in/application extension.
Basic operation of the control package includes parsing of multicast
control commands (traffic on port 200), their interpretation and
generation of appropriate responses. Before passing in any control
commands, the application must activate the package by feeding it the
root key which will be used to derive multicast key encryption key (McKEKey).
This McKEKey is then ued to derive network and application session keys
when the multicast address is conveyed to the device using
MCAST_GROUP_SETUP request. The root key in case of LoRaWAN 1.0.X is a
special out-of-band pre-provisioned key. For v1.1.X AppKey is used as
root key for this plugin too.
The application is responsible for the transmission of responses at port
200 which are generated by the plug-in.

Commit:
Adding Fragmentation control plug-in

This is an implementation of LoRaWAN Fragmentation Control Block
Transport Specification v1.0.0. It can be used in a variety of ways,
e.g., receiving a large payload chopped into smaller fragments. A
specific example where this plug-in can be used is the firmware update.

The plug-in involves three components:

FragmentationControlPackage -
 Helps the application in parsing, interpreting, and responding to control commands.
FragAssembler -
 Keeps track of fragments received, fragments lost and performs FEC
harnessing redundancy packets.
FragBDWrapper -
 A wrapper class for a block device

ciarmcom commented:
@hasnainvirk, thank you for your changes.
@AnttiKauppila @janjongboom @ARMmbed/mbed-os-wan @ARMmbed/mbed-os-maintainers please review.

Commit:
Astyle fixes for LoRaWAN plugins

Left over style issues for LoRaWAN plugins and sessions are resolved in
this commit.

AnttiKauppila commented on features/lorawan/mbed_lib.json:
```diff
@@ -92,6 +92,22 @@
+            "value": 37
+        },
+        "gps-epoch-tai-minus-utc": {
+            "help": " GPS epoch TAI minus UTC time, default 19 (as of 2018)",
```
Not needed

AnttiKauppila commented on features/lorawan/system/lorawan_data_structures.h:
```diff
@@ -69,6 +69,43 @@ typedef uint32_t lorawan_time_t;
+
+inline uint8_t *write_four_bytes(uint32_t value, uint8_t *ptr)
+{
+    if (!ptr) {
```
No implementation in header files

hasnainvirk replied:
It's an inline function. The problem is, it's harder to find a central location which accessible by every component. The idea was to make it available as a global helper without limiting its scope. Maybe we need to rethink.  If I remember correctly, there are other examples of inline function implementations in the headers in Mbed OS.

AnttiKauppila replied:
Yes there are, but it does not mean it is a correct way. We should also clean those going forward.
You will need something like common_helpers.cpp to put implementation into.

AnttiKauppila reviewed (changes requested):
Internal review needed before this can be merged

Commit:
Moving byte access helpers to source

Very simple byte access helper functions 'write_four_bytes'
and read_four_bytes()' are moved to their own source file rather
than placing them in lorawan_data_structures.h.

Commit:
Adding missing doc for multicast APIs in LoRaMac

API Docs are added for 'update_multicast_addr_register(...)' and
'check_multicast_params(...)' which was missing from previous commits.

Commit:
Adding network assisted GPS time acquisition

LoRaWANTimer class is responsible for providing time base for the
LoRaWAN stack in Mbed OS. It derives its core monotonic clock from the
EventQueue which is essentially a millisecond ticker or equivalent (in
tickless mode). LoRaWAN network can assist the device for device level
time synchronization using ClockSynchronization protocol (for v1.0.2) or
by replying to DevTimeReq MAC command (for v1.0.3 and above). The time
base used for device synchronization using above methods is purely
based on GPS epoch and needs no conversion to UTC or TAI etc (no
adjustment for leap seconds either).
Therefore we have extended the LoRaWANTimer APIs to account for GPS time
base if assisted by the network.
get_gps_time() is the API to use to acquire GPS time base from the
stack. If the network never assisted the device with a time stamp, a
value of zero is returned which means GPS time base is not available.
set_gps_time(time) API is used to set a given GPS time-stamp when the
network sends it.
These APIs are hooked via LoRaMAC to LoRaWANInterface which exposes the
access to the application.

Commit:
Exposing GPS/UTC time APIs

Provides hooks for the internal APIs handling GPS time base for device
synchronization.
In addition to that 'set_system_time_utc(...)' API is added which have no
roots in the stack. Its just a helper for the application developer to
acquire UTC time base given the difference between TAI and UTC time
bases.

Commit:
Using GPS time base for plug-ins

Conversion to UTC time base and utilization is not necessitated by the
LoRa Alliance and the Network servers are expected to use GPS time base
without leap second conversion at least for device synchronization.
Hence we modify the code to account for only GPS time base.

Commit:
Making COUNT_SET_BITS a function

Using macros to implement logic is discouraged. Changing COUNT_SET_BITS
to a helper function count_set_bits(...).

Commit:
Removing configurations for time conversions

We don't need to convert for UTC from GPS time, hence we don't need
configurations for time base differences.

Commit:
Adding DR alongwith Frequency for RX2 config

An API was introduced to change RX2 frequency on demand. However, the
config should be coupled with a fixed on-demand data rate as well.
We have modified the API to account for data rate too.

0xc0170 commented on features/lorawan/plugins/FragBDWrapper.h:
```diff
@@ -0,0 +1,91 @@
+/*
+ * PackageLicenseDeclared: Apache-2.0
```
Not SPDX-License-Identifier instead as in other files, I Can see just two here using this identifier instead

hasnainvirk replied:
@0xc0170 FragAssembler is based on third party code. But yeah, you are right, FragBDWrapper should have SPDX-License-Identifier. Will fix that. 
For testing: Unit tests will be added to the PR once https://github.com/ARMmbed/mbed-os/pull/9219 gets merged. After #9219, we will do a doc update to this PR plus unit tests.  
For protocol implementation testing, we wrote an Application Server that talks to Public TTN LoRaWAN network . It can be found here https://github.com/hasnainvirk/lorawan-ttn-as-test. 
At the moment there is no real life implementation for these protocols (plug-ins). We are the first.  That's why we had to come up with this testing solution. The AS is written in golang, so I am not sure how we can integrate it in CI. However, we would definitely like to see some tests around this.

0xc0170 reviewed (commented):
No tests being added?
